#pragma once
#include "const.h"
#include "Singleton.h"
#include "ConfigMgr.h"
#include "message.grpc.pb.h"
#include "message.pb.h"
#include <grpcpp/grpcpp.h>
#include <queue>
#include <condition_variable>

using grpc::Channel;
using grpc::Status;
using grpc::ClientContext;

using message::GetChatServerReq;
using message::GetChatServerRsp;
using message::LoginRsp;
using message::LoginReq;
using message::StatusService;

class StatusConPool {
public:
	StatusConPool(size_t poolSize, std::string host, string port) 
		:poolSize(poolSize), host_(host), port_(port), b_stop(false){
		for (size_t i = 0; i < poolSize; ++i) {
			std::shared_ptr<Channel> channel = grpc::CreateChannel(host_ + ":" + port_, 
				grpc::InsecureChannelCredentials());
			connections_.push(StatusService::NewStub(channel));
		}
	}

	~StatusConPool() {
		std::lock_guard<std::mutex> lock(mutex_);
		Close();
		while (!connections_.empty()) {
			connections_.pop();
		}
	}

	std::unique_ptr<StatusService::Stub> getConnection() {
		std::unique_lock<std::mutex> lock(mutex_);
		cond_.wait(lock, [this] {
			if (b_stop) {
				return true;
			}
			return !connections_.empty();
		});
		if (b_stop) {
			return nullptr;
		}
		auto context = std::move(connections_.front());
		connections_.pop();
		return context;
	}	

	void returnConnection(std::unique_ptr<StatusService::Stub> context) {
		std::lock_guard<std::mutex> lock(mutex_);
		if (b_stop) {
			return;
		}
		connections_.push(std::move(context));
		cond_.notify_one();
	}	

	void Close() {
		b_stop = true;	
		cond_.notify_all();
	}	

private:
	atomic<bool> b_stop;
	size_t poolSize;
	std::string host_;
	std::string port_;
	std::mutex mutex_;
	std::condition_variable cond_;
	std::queue<std::unique_ptr<StatusService::Stub>> connections_;
};

class StatusGrpcClient : public Singleton<StatusGrpcClient> {
	friend class Singleton<StatusGrpcClient>;
public:
	~StatusGrpcClient() {}
	GetChatServerRsp GetChatServer(int uid);
	LoginRsp Login(int uid, std::string token);

private:
	StatusGrpcClient();
	std::unique_ptr<StatusConPool> pool_;
};